semgrep/semgrep/rule.py

```diff
@@ -248,6 +250,47 @@ def has_runnable_rule(d: Dict[str, Any]) -> bool:
 
         return has_runnable_rule(self._raw)
 
+    @property
+    def formula_string(self) -> str:
+        """
+        Used to calculate a pattern based ID, works through DFS of all
+        PATTERN KEYS, and additionally, if the key is 'join' we include the 'on' field also
+        """
+
+        def get_subrules(raw: Union[AnyStr, Dict, List]) -> str:
+            patterns = ""
+            if isinstance(raw, str):
+                return raw
+            elif isinstance(raw, dict):
+                for k in RuleValidation.PATTERN_KEYS:
```

Comment:
Do we feel confident in the order of this being stable? This is part of how we navigate the tree, so maybe we should sort it here, but it's also a constant, so maybe we should just manually sort it or leave a comment?

Reply:
I think we should sort it ourselves. After 3.7 sets should by default be ordered but can't hurt to sort the 10 elements here in case changes are made to PATTERN_KEYS